### PR TITLE
Deny production runs of example.org trust domains

### PR DESCRIPTION
**Pull request by kfox1111**
This pr adds some tests to ensure that if the user is deploying for production that they set appropriate settings.

fixes: https://github.com/spiffe/helm-charts/issues/195

### CODE REVIEW

**Review by marcofranssen — COMMENTED** [review]
Do we really want to enforce people to override? If they are just doing a quick install for testing, shouldn't that be possible with the default values?

If we want people to explicitly set those it would be better to remove the defaults.

This feels a bit like a unnecessary additional config value.

**Comment by kfox1111** [comment]
> Do we really want to enforce people to override? If they are just doing a quick install for testing, shouldn't that be possible with the default values?

That's the point of the pr. If you just helm install the chart, it just works with example.org for a quick install for testing as it does today.

If you say explicitly, your trying to deploy for production by adding examples/production/values.yaml, then it forces you to specify the things you should be specifying. No one should be deploying a production spire as example.org? The pr helps you find all the things you need to override to get a valid production site by enforcing you set them.
 
> If we want people to explicitly set those it would be better to remove the defaults.

That would make the, "just kick the tires" install much harder. But I could get behind that if there is agreement that is easier. It would be less complicated in the chart, just harder to get an example going. We could put the example values in examples/testing-deployment/values.yaml or something and then have people need to use that for a test deployment?

> This feels a bit like a unnecessary additional config value.

Necessary or unnecessary depends on the usability path question above.

File: charts/spire/templates/_spire-lib.tpl
Comment (kfox1111) [reply charts/spire/templates/_spire-lib.tpl:61]: Yeah... A usage documentation item would be good. Will add.

Ordering of the yaml doesn't matter. useage is like:
```
{{- include "spire-lib.strict-check" (list . "trustDomain must be set" (eq (include "spire-lib.trust-domain" .) "example.org"))}}
```

where the arguments are specified in a list so wont be reordered, but takes up less space then doing it with dictionary passing.

args are root, an error to display, and a boolean on if to fail and display when true

Comment (kfox1111) [reply charts/spire/templates/_spire-lib.tpl:61]: actually, the arguments are kind of described here:
https://github.com/spiffe/helm-charts/pull/229/files#diff-789477a4f2a924ca09b7f80a469fc8827159a3d3d8a6281c37018dc16b60c094R55-R56

There any particular wording changes that might make it more understandable?

Comment (kfox1111) [reply charts/spire/templates/_spire-lib.tpl:61]: > * What is `index . 2`?
> * What is `index . 1`?

Oh. Maybe I'm not understanding the question either....
```index thing 1``` is what might be done in other languages like ```thing[1]``` or sometimes ```thing.1```. its looking up the second argument passed to the macro in a list. (0 based indexing)

Comment (marcofranssen) [reply charts/spire/templates/_spire-lib.tpl:63]: I'm also having some difficulty processing this one

```
{{- include "spire-lib.strict-check" (list . "trustDomain must be set" (eq (include "spire-lib.trust-domain" .) "example.org"))}}
```

I think even though this might be less syntax a dict could make it more descriptive and understandable. After reading this example 3 times I'm still not sure I get it.

File: charts/spire/charts/spire-agent/templates/configmap.yaml
Comment (edwbuck) [inline charts/spire/charts/spire-agent/templates/configmap.yaml:1]: We should not provide a default cluster name or trust domain.

Out of all of the settings, these should be specified by the user 100% of the time.

**Review by edwbuck — DISMISSED** [review]

File: charts/spire/templates/_spire-lib.tpl
Comment (marcofranssen) [reply charts/spire/templates/_spire-lib.tpl:63]: Ah got it now. Probably a good suggestion to assign them to variables that makes it more understanable indeed.

I would suggest following names.

```
{{ $root := index 0 }}
{{ $message := index 1 }}
{{ $condition := index 2 }}
```

Comment (kfox1111) [reply charts/spire/templates/_spire-lib.tpl]: Done

File: examples/production/values.yaml
Comment (kfox1111) [reply examples/production/values.yaml:3]: Done

File: charts/spire/charts/spire-agent/templates/configmap.yaml
Comment (kfox1111) [reply charts/spire/charts/spire-agent/templates/configmap.yaml:1]: What about macro's suggestion of still keeping the ability to easily kick the tires, but turn it off by default?

File: examples/production/values.yaml
Comment (kfox1111) [reply examples/production/values.yaml:3]: Actually, it looks like a bigger lift to make sure all the tests are then specifying the newly required config too. I can do that, but we should decide that's the way we want to go, first.

File: charts/spire/charts/spire-agent/templates/configmap.yaml
Comment (edwbuck) [reply charts/spire/charts/spire-agent/templates/configmap.yaml:1]: I'm willing to be flexible, but I really don't like the precedent that we either have already set, or are committing to setting.

If a installation using our Helm charts isn't production quality, then we should fix it to be production quality, adjust what we decide is production quality, or adjust our documentation to indicate that initially the installation won't be production quality, but after you set a specific property, it will be?

Personally I don't care which one is the path forward; and in the interest of getting some movement, and @marcofranssen call that we easily "kick the tires, but turn (I'm not sure if it is "production quality" or "easy non-production prototype installations", I'm going to withdraw my objection and clear this for the other maintainers to move in whatever direction they deem fit.

**Review by kfox1111 — COMMENTED** [review]
Update to include new global.

Also, set default back to false until we determine that tests should be rewritten

**Comment by faisal-memon** [comment]
> Something like productionChecks for the value instead.

That sounds good to me.

File: charts/spire/values.yaml
Comment (faisal-memon) [inline charts/spire/values.yaml:27]:
```suggestion
    ## @param global.spire.productionChecks Check values, such as trustDomain, are suitable for production deployment
```

Comment (faisal-memon) [reply charts/spire/values.yaml:27]: I like the description to say what enabling the flag does

**Review by faisal-memon — APPROVED** [review]

File: charts/spire/charts/spire-server/values.yaml
Comment (marcofranssen) [inline charts/spire/charts/spire-server/values.yaml:228]: https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes

Which country does ARPA represent?

File: charts/spire/values.yaml
Comment (marcofranssen) [inline charts/spire/values.yaml:28]: Shall we call this `strictMode`? that is also something used quite commonly in Javascript, React, Typescript world.

File: charts/spire/charts/spire-server/values.yaml
Comment (kfox1111) [reply charts/spire/charts/spire-server/values.yaml:228]: https://www.digicert.com/kb/ssl-certificate-country-codes.htm

ARPA should be valid, but shouldn't be used in any org I think with spire. so easy to check if its been set or not by the end user?

File: charts/spire/values.yaml
Comment (kfox1111) [reply charts/spire/values.yaml:28]: @faisal-memon ?

File: examples/production/values.yaml
Comment (marcofranssen) [inline examples/production/values.yaml:6]:
```suggestion
   strictMode: true
```

As suggested earlier, maybe call it strictMode. Shouldn't the value be true for this test?

File: charts/spire/templates/_spire-lib.tpl
Comment (marcofranssen) [inline charts/spire/templates/_spire-lib.tpl:119]:
```suggestion
{{- if (dig "spire" "strictMode" false $root.Values.global) }}
```

File: charts/spire/templates/_spire-lib.tpl
Comment (marcofranssen) [inline charts/spire/templates/_spire-lib.tpl:115]:
```suggestion
{{- define "spire-lib.check-strict-mode" }}
```

File: charts/spire/templates/_spire-lib.tpl
Comment (marcofranssen) [inline charts/spire/templates/_spire-lib.tpl:113]:
```suggestion
if strictMode is enabled and the boolean is true
```

File: charts/spire/charts/spire-server/templates/configmap.yaml
Comment (marcofranssen) [inline charts/spire/charts/spire-server/templates/configmap.yaml:6]:
```suggestion
{{- include "spire-lib.check-strict-mode" (list . "clusterName must be set" (eq (include "spire-lib.cluster-name" .) "example-cluster"))}}
{{- include "spire-lib.check-strict-mode" (list . "trustDomain must be set" (eq (include "spire-lib.trust-domain" .) "example.org"))}}
{{- include "spire-lib.check-strict-mode" (list . "jwtIssuer must be set" (eq (include "spire-lib.jwt-issuer" .) "https://oidc-discovery.example.org"))}}
{{- include "spire-lib.check-strict-mode" (list . "ca_subject.county must be set" (eq .Values.ca_subject.country "ARPA"))}}
{{- include "spire-lib.check-strict-mode" (list . "ca_subject.organization must be set" (eq .Values.ca_subject.organization "Example"))}}
{{- include "spire-lib.check-strict-mode" (list . "ca_subject.common_name must be set" (eq .Values.ca_subject.common_name "example.org"))}}
```

File: charts/spire/charts/spire-agent/templates/configmap.yaml
Comment (marcofranssen) [inline charts/spire/charts/spire-agent/templates/configmap.yaml:2]:
```suggestion
{{- include "spire-lib.check-strict-mode" (list . "clusterName must be set" (eq (include "spire-lib.cluster-name" .) "example-cluster"))}}
{{- include "spire-lib.check-strict-mode" (list . "trustDomain must be set" (eq (include "spire-lib.trust-domain" .) "example.org"))}}
```

File: charts/spire/charts/spiffe-oidc-discovery-provider/templates/configmap.yaml
Comment (marcofranssen) [inline charts/spire/charts/spiffe-oidc-discovery-provider/templates/configmap.yaml:2]:
```suggestion
{{- include "spire-lib.check-strict-mode" (list . "trustDomain must be set" (eq (include "spire-lib.trust-domain" .) "example.org"))}}
{{- include "spire-lib.check-strict-mode" (list . "jwtIssuer must be set" (eq (include "spire-lib.jwt-issuer" .) "https://oidc-discovery.example.org"))}}
```

File: charts/spire/README.md
Comment (marcofranssen) [inline charts/spire/README.md:101]:
```suggestion
| `global.spire.strictMode`         | Check values, such as trustDomain, are suitable for production deployment                                  | `false`                      |
```

File: examples/production/values.yaml
Comment (kfox1111) [reply examples/production/values.yaml:6]: Good catch. broke in the last rename I guess.

**Review by marcofranssen — CHANGES_REQUESTED** [review]

**Review by marcofranssen — APPROVED** [review]
LGTM 🚀

**Pull request merged**
